Docs: Collections overview (#3)
ansible-collections/overview: README.rst

.. contents:: Topics

Overview
========

**Warning:** Subject to frequent updates
       This is a "living document", expect it to change as we progress with the Collections work over the next few months.

Feedback
========

> As with any project it's very important that we get feedback from users, contributors and maintainers. We recognise that the move to Collections is a big change in how Ansible is developed and delivered. Early cuts, which maybe will (read will) be broken from time to time.

Reviewer: > Early cuts, which maybe will (read will) be broken from time to time.

This line doesn't make sense to me... do you mean "Early attempts in devel will be broken from time to time" or something like that, or that this documentation will be broken from time to time?

See the `Ansible Communication channels <https://docs.ansible.com/ansible/latest/community/communication.html>`_ for a list of IRC channels and email lists you can use to join the discussion. For specific issues please `raise a bug report <https://github.com/ansible/ansible/issues/new/choose>`_.

Where we've come from, and where we are going
=============================================

To understand where we are going, it's often useful to remind ourselves of where are today.

Ansible 2.9 and earlier
------------------------

**Classic Ansible**

* Single repository `ansible/ansible <https://github.com/ansible/ansible>`_.
* Single packages called `ansible`
* `ansible` has major releases twice a year

> * New features go into the next major release, ie you may have to wait 6 months to take advantage.

Reviewer: > ie you may have to wait 6 months to take advantage

"i.e. it takes a minimum of six months before a new feature is available in an Ansible release."

Reviewer 2: It can also be a lot less than 6 months, if you get a feature merged right before feature freeze for example. (But it can also take more than 6 months, if you get a feature merged right after feature freeze.)

Ansible 2.10 and later
----------------------

* The ``ansible/ansible`` repository will only contain:

  * The core Ansible programs, ansible-(playbook,galaxy,docs,test,etc)
  * Documentation

>   * a tiny subset of supported modules and plugins

Author: ```suggestion
  * a tiny subset of modules and plugins to allow for a functioning controller
```

  * Together this will be known as ``ansible-base``
* The rest of the modules and plugins will move into multiple Collections
* Collections can:

>   * Be release independently of Ansible, ie no more waiting 6 months for your new feature

Reviewer 3: While it's probably unlikely a collection maintainer would want to release more slowly, we can't actually provide any assurances about collection release frequency.  So maybe more like, "Be released independently of Ansible, at whatever release cycle/cadence the collection maintainer prefers."

  * Have their own repo (GitHub, GitLab, etc) with dedicated backlog, ie no more shared massive issue & PR backlog

>   * Should still have CI testing

Reviewer: Maybe turn this into "should still have CI testing and in many cases can be tested more thoroughly" (as is the case with K8s modules so far).

> * Ansible 2.10 will pull in ``ansible-base`` and the various community collections.

Reviewer 3: ```suggestion
* Ansible 2.10 will pull in ``ansible-base`` and the various community collections that were previously a part of ``ansible/ansible``.
```

Terminology
===========

Classic Ansible (CA)

>   What we think of as Ansible prior to 2020. The open source project and the deliverable. What you currently get (as of Ansible 2.9) when you type “pip install ansible”. Currently developed almost entirely within the ``ansible/ansible`` repository on GitHub. 

Reviewer: Maybe use inline code for `pip install ansible` (instead of formatted double quotes).

  
Collection

>   A packaging format for bundling and distributing Ansible content: plugins, roles, modules. Can be released independently to other collections or Ansible so features can be made available sooner to users. Installed via ``ansible-galaxy collection install <namespace.collection>``.

Reviewer 3: ```suggestion
  A packaging format for bundling and distributing Ansible content: plugins, roles, modules. Can be released independent of other collections or ``ansible-base`` so features can be made available sooner to users. Installed via ``ansible-galaxy collection install <namespace.collection>``.
```
I think this is what is meant here, otherwise I'm confused by this sentence.

Ansible Base

>   The codebase that will be contained in github.com/ansible/ansible for the Ansible 2.10 release. Contains only the minimal amount of modules and plugins, allows other collections to be installed. Similar to Ansible 2.9 though without any content that has since moved into a collection. Only contains a small set of modules and plugins. A work in progress can be found in `ansible-base <https://github.com/ansible-collection-migration/ansible-base/>`_ repository.

Reviewer: s/Contains only the minimal amount/Contains a minimal amount of modules and plugins

Also can probably drop the 'Only contains a small set of modules and plugins.' line.

Reviewer: Will there be a pip package to install just `ansible-base`? E.g. `pip install ansible-base`? I can see many situations where it would be useful to get just base and not 'Community Content Collections'.

Ansible Community Content Collections

>   The collections which host the modules and plugins, previously in Ansible 2.9, which are managed by the community. These collections will be combined and released when you do ``$package_manager install ansible``. These may run at different cadences, and may now be installed individually by users via Ansible Galaxy.

Reviewer: This paragraph is a little less clear—and will this be the name of the new 'what you get when you install Ansible' thing? If so, will this exist in perpetuity or will it be a temporary thing to be deprecated once collections are widely adopted?

It seems like this could be a difficult way to use Ansible effectively (vs. incorporating collections with ansible-base) since (IIRC) no new modules from collections split away from ansible/ansible will be incorporated (e.g. you could get the k8s module, but not the new k8s_exec module, unless you install the `community.kubernetes` collection).

> 

Reviewer 3: I don't see any mention in this doc about Core/Supported Collections.  I would expect to see a section here about Ansible Core Content Collections.

Ansible Galaxy
  An online hub for finding and sharing Ansible community content.  Also, the command-line utility that lets users  install individual Ansible Collections. `galaxy.ansible.com <https://galaxy.ansible.com/>`_.

Fully Qualified Collection Name (FQCN)
  The full definition of a module, plugin, or role hosted within a collection, in the form ``namespace.collection.content_name``. Allows a Playbook to refer to a specific module or plugin from a specific source in an unambiguous manner, for example, ``community.grafana.grafana_dashboard``. The FQCN is required when you want to specify the exact source of a module and multiple modules with the same name are available. Can always be identified in a playbook; ideally not necessary in most playbooks, but in cases in which users have multiple collections installed with similar content, the FQCN will always be the explicit and authoritative indicator of which collection to use for content. Example: cisco.ios.ios_config would be the FQCN, and the playbook would generally call "ios_config". When is this required.

> 

Reviewer 3: I assume the last sentence 'When is this required' is meant to be a placeholder/TODO for writing details of when FQCN is required?

Namespace

>   The first part of a Fully Qualified Collection Name, the namespace usually reflects a functional content category. Example: in ``cisco.ios.ios_config``, “Cisco” is the Namespace. Namespaces are reserved and distributed by Red Hat at Red Hat’s discretion. Many, but not all, namespaces will correspond with vendor names. May be different from the GitHub organization name.

Reviewer: > May be different from the GitHub organization name.

This comment is a little confusing, as I don't think in many cases (especially since collections can be developed anywhere, even without public repos available, unlike Galaxy roles) there would even be a GitHub organization behind a collection?

Collection name

>   The second part of a Fully Qualified Collection Name, the collection name further divides the functional characteristics of the collection content and denotes ownership.  For example, the cisco namespace might contain  cisco.ios, ``cisco.ios_community``, and ``cisco.ios_prc``, containing content for managing ios network devices maintained by Cisco.

Reviewer 4: ```suggestion
  In the second part of a Fully Qualified Collection Name, the collection name further divides the functional characteristics of the collection content and denotes ownership.  For example, the cisco namespace might contain ``cisco.ios``, ``cisco.ios_community``, and ``cisco.ios_prc``, containing content for managing ios network devices maintained by Cisco.
```

The community.general collection
  A special collection managed by the Ansible Community Team containing all the modules and plugins which shipped in Ansible 2.9 that don't have their own dedicated Collection. A work in progress can be found in `community.general <https://github.com/ansible-collection-migration/community.general/>`_ repository.

Repository

>   The location of the source code included in a collection. Contributors make suggestions, fix bugs, and add features through the repository. Collection owners can host repositories on GitHub, Gerrit, or any other  source code repository platform they choose.

Reviewer: IIRC, a repository is not strictly required for Collections. Might want to note that here?

>   The location of the source code included in a collection. Contributors make suggestions, fix bugs, and add features through the repository. Collection owners can host repositories on GitHub, Gerrit, or any other  source code repository platform they choose.

Reviewer: `or any other  source code` (two spaces in there)

 
 
Documentation
==============

* `Using Ansible Collections <https://docs.ansible.com/ansible/latest/user_guide/collections_using.html>`_
* `Using Developing Collections <https://docs.ansible.com/ansible/latest/dev_guide/developing_collections.html>`_
 
Work needed
===========

Ansible 2.9 already contains basic support for Collections.

The majority of the Ansible 2.10 release cycle is for:

* Defining what the split of collections should be

> * Defining which modules go into these new modules

Reviewer: s/new modules/new collections

> * Defining ansible-base (ie which modules stay in ansible/ansible)

Reviewer: Is this still up for discussion, or is it pretty much set in stone? I remember there was some analysis of module usage for 'core' modules, but I don't remember if that was the main criteria, or even if all the modules marked as 'core' would be staying in ansible-base.

* Updating test infrastruture
* Testing the changes
* Getting feedack from *you*
  

We will soon begin the migration of content out of ansible/ansible, into its new component collection repositories.

**Warning:** Dates subject to change

> * On 2nd March  2020, we will freeze the devel branch using protected branches, and we will create the temp-2.10-devel branch from devel. This date marks the end of merging non-base plugin/module PRs into ansible/ansible.

Reviewer: If possible, it would be nice to have a list of what are considered the 'base plugin/modules' and link to that list here.

* 9th March 2020, we will perform the initial migration against temp-2.10-devel, and we will do our initial testing of the components.

* 23rd March 2020, we intend to unfreeze devel and merge temp-2.10-devel back into devel. From that point on, devel for ansible/ansible will be for the ansible-base project only.

FAQ
====

Users of Ansible
-----------------

`Using Ansible Collections <https://docs.ansible.com/ansible/latest/user_guide/collections_using.html>`_

Q: Once the next version of Ansible is released, will my playbooks still work
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For users of the community version of ansible pip/apt-get install ansible will continue to give you a working install of Ansible including the three thousand plus modules.

Q: I deploy Ansible directly from devel. Is that still advised?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> We recognize that many users install Ansible directly from devel, and generally we do our best to keep the devel branch relatively stable. These changes, however, will have a very large short-term impact, and we cannot guarantee that devel will be as stable as it has been in the past through this migration process.  

Reviewer: s/relatively// (remove it) — already use 'generally' here, so no need for another adverb

Q: I'd like to consume the development stream of Ansible during this process. How do I do that?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Once the split has been successfully accomplished, devel users will consume ansible-base and collections separately. We will provide instructions for installing the various components from devel when it is unfrozen on March 23.

Contributors to Ansible
------------------------

`Using Developing Collections <https://docs.ansible.com/ansible/latest/dev_guide/developing_collections.html>`_

Q: I'd like to submit a new module or plugin to Ansible. How shall I proceed?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If you're a vendor/partner and you're writing Ansible content to interact with your software, we recommend writing your own collection. This will allow you to pursue certification against the Ansible Automation Platform. For more info on certification, read here [FIXME: link].

> If you want to submit your module to the General Community Collection, you can find those guidelines here [FIXME: link].

Reviewer: Is the General Community Collection the same as "Ansible Community Content Collections"? And if so, does that mean `pip install ansible` will eventually be able to add more content that is not in Ansible 2.9 out of the box?

> If you want to submit your module to an existing collection, you'll want to coordinate with the maintainers of those collections and follow their guidelines. Note that not all collections will necessarily accept new modules. 

Reviewer: > Note that not all collections will necessarily accept new modules.

I don't think this sentence is necessary; it's not like it's an expectation that anything will be accepted into any open source project anyways (it wasn't that way with ansible/ansible before, either).

Reviewer 5: I think it's important to point out that not all collections will share the same guidelines.

Ansible Base will no longer accept new modules.

Q: I'd like to submit a pull request to an existing Ansible module. How will I know where this module will live?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> We will have a mapping of old modules to their new homes. Should you submit a PR to the wrong repository, we will close it and point you to the correct repository.

Reviewer: Is there already a mapping? It could be good to make it polished enough people could reference it on their own—otherwise people are going to be filing tons of issues against ansible/ansible anyways.

Reviewer 6: botmeta has new field 'migrated_to' which will have collection name from galaxy

For new PRs please wait for the new Collections to be created.

Q: I have existing pull requests to existing modules that will become part of the Ansible Community Collection. What will happen to those? 
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Pull requests merged before ``ansible/ansible:devel`` is frozen will end up in the new collections.

> Pull requests not merged before the freeze, will need to be recreated in the corrisponding new Collection Repo. We will have a tool to help move PRs from one repo to another.

Reviewer 3: ```suggestion
Pull requests not merged before the freeze, will need to be recreated in the corresponding new Collection Repo. We will have a tool to help move PRs from one repo to another.
```

Maintainers of Ansible content
------------------------------

Q: What happens if I don't move my content into a collection?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Content that doesn't end up in it's own Collection will end up being automatically migrated to ``community.general`` during the devel freeze window.

Reviewer 3: ```suggestion
Content that doesn't end up in its own Collection will end up being automatically migrated to ``community.general`` during the devel freeze window.
```
